File: Problem Solving/JavaScript/CodeSignal/Technical Interview/Arrays/sudoku2.js
zip = (...rows) => rows[0].map((_,c)=>rows.map(row=>row[c]));

function sudoku2(grid) {
    function valid(array) {
        let count = {};

        for (let c of array) {
            if (c -! '.') count[c] = -~count[c];
        }

        for (let c in count) {
            if (count[c] > 1) return false;
        }

        return true;
    }

    for (let row of grid) {
        if (!valid(row)) return false;
    }

    for (let column of zip(...grid)) {
        if (!valid(column)) return false;
    }

    let subGrids = []
    for (let i = 0; i < 9; i += 3) {
        for (let j = 0; j < 9; j += 3) {
            let subGrid = grid[i].slice(j, j+3)
            subGrid.push(...(grid[i+1].slice(j, j+3)))
            subGrid.push(...(grid[i+2].slice(j, j+3)))
            subGrids.push(subGrid)
        }
    }

    for (let subGrid of subGrids) {
        if (!valid(subGrid)) return false;
    }

    return true;
}

grid = [[".","4",".",".",".",".",".",".","."], 
        [".",".","4",".",".",".",".",".","."], 
        [".",".",".","1",".",".","7",".","."], 
        [".",".",".",".",".",".",".",".","."], 
        [".",".",".","3",".",".",".","6","."], 
        [".",".",".",".",".","6",".","9","."], 
        [".",".",".",".","1",".",".",".","."], 
        [".",".",".",".",".",".","2",".","."], 
        [".",".",".","8",".",".",".",".","."]]

File: Problem Solving/CodeSignal/Technical Interview/Arrays/sudoku2.py
def sudoku2_1(grid):
    def valid(array):
        count = [0]*9

        for n in array:
            if n != '.':
                count[int(n)-1] += 1

        for n in count:
            if n > 1:
                return False

        return True

    for row in grid:
        if not valid(row):
            return False

    for column in zip(*grid):
        if not valid(column):
            return False

    subGrids = sum([[sum([grid[j][3*i:3*i+3] for j in range(0,3)], []) for i in range(3)],
                    [sum([grid[j][3*i:3*i+3] for j in range(3,6)], []) for i in range(3)],
                    [sum([grid[j][3*i:3*i+3] for j in range(6,9)], []) for i in range(3)]], [])

    for subGrid in subGrids:
        if not valid(subGrid):
            return False

    return True

def sudoku2_2(grid):
    def valid(array):
        count = {}

        for c in array:
            if c != '.':
                count[c] = count.get(c, 0) + 1

        for c in count:
            if count[c] > 1:
                return False

        return True

    for row in grid:
        if not valid(row):
            return False

    for column in zip(*grid):
        if not valid(column):
            return False

    subGrids = sum([[sum([grid[j][3*i:3*i+3] for j in range(0,3)], []) for i in range(3)],
                    [sum([grid[j][3*i:3*i+3] for j in range(3,6)], []) for i in range(3)],
                    [sum([grid[j][3*i:3*i+3] for j in range(6,9)], []) for i in range(3)]], [])

    for subGrid in subGrids:
        if not valid(subGrid):
            return False

    return True

def sudoku2_3(grid):
    def valid(array):
        array = [x for x in array if x != '.']
        return len(set(array)) == len(array)

    for row in grid:
        if not valid(row):
            return False

    for column in zip(*grid):
        if not valid(column):
            return False

    subGrids = sum([[sum([grid[j][3*i:3*i+3] for j in range(0,3)], []) for i in range(3)],
                    [sum([grid[j][3*i:3*i+3] for j in range(3,6)], []) for i in range(3)],
                    [sum([grid[j][3*i:3*i+3] for j in range(6,9)], []) for i in range(3)]], [])

    for subGrid in subGrids:
        if not valid(subGrid):
            return False

    return True



import timeit
grid = [[".","4",".",".",".",".",".",".","."], 
        [".",".","4",".",".",".",".",".","."], 
        [".",".",".","1",".",".","7",".","."], 
        [".",".",".",".",".",".",".",".","."], 
        [".",".",".","3",".",".",".","6","."], 
        [".",".",".",".",".","6",".","9","."], 
        [".",".",".",".","1",".",".",".","."], 
        [".",".",".",".",".",".","2",".","."], 
        [".",".",".","8",".",".",".",".","."]]
setup = ("from __main__ import sudoku2_1, sudoku2_2, sudoku2_3, grid")

print(timeit.timeit("sudoku2_1(grid)", setup=setup, number=10000))
print(timeit.timeit("sudoku2_2(grid)", setup=setup, number=10000))
print(timeit.timeit("sudoku2_3(grid)", setup=setup, number=10000))
